public/app/plugins/panel/canvas/editor/TreeNavigationEditor.tsx

```diff
@@ -0,0 +1,218 @@
+import { PanelOptions } from '../models.gen';
+import { getTreeData, TreeElement } from '../tree';
+import { LayerActionID } from '../types';
+import { doSelect } from '../utils';
+
+import { TreeViewEditorProps } from './treeViewEditor';
+
+type Props = StandardEditorProps<any, TreeViewEditorProps, PanelOptions>;
+
+type State = {
+  treeData: TreeElement[];
+  autoExpandParent: boolean;
+  expandedKeys: Key[];
+};
+
+export class TreeNavigationEditor extends PureComponent<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      treeData: getTreeData(props.item?.settings?.scene.root),
+      autoExpandParent: true,
+      expandedKeys: [],
+    };
+  }
+
+  globalCSS = getGlobalStyles(config.theme);
+  settings = this.props.item.settings;
+  theme = getTheme();
+  styles = getStyles(this.theme);
+  rootElements = this.props.item?.settings?.scene.root.elements;
+  selectedBgColor = this.theme.colors.border1;
+
+  onSelect = (selectedKeys: Key[], info: any) => {
+    doSelect(this.settings, info.node.dataRef);
+  };
+
+  allowDrop = (info: any) => {
+    if (!info.dropNode.children) {
+      if (info.dropPosition === 0) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  onDrop = (info: any) => {
+    const destKey = info.node.key;
+    const srcKey = info.dragNode.key;
+    const destPos = info.node.pos.split('-');
+    const destPosition = info.dropPosition - Number(destPos[destPos.length - 1]);
+
+    const srcEl = info.dragNode.dataRef;
+    const destEl = info.node.dataRef;
+
+    const loop = (data: TreeElement[], key: number, callback: any) => {
+      data.forEach((item, index, arr) => {
+        if (item.key === key) {
+          callback(item, index, arr);
+          return;
+        }
+        if (item.children) {
+          loop(item.children, key, callback);
+        }
+      });
+    };
+    const data = [...this.state.treeData];
+
+    // Find dragObject
+    let srcElement: TreeElement;
+    loop(data, srcKey, (item: TreeElement, index: number, arr: TreeElement[]) => {
+      arr.splice(index, 1);
+      srcElement = item;
+    });
+
+    if (destPosition === 0) {
+      // Drop on the content
+      loop(data, destKey, (item: TreeElement) => {
+        item.children = item.children || [];
+        item.children.unshift(srcElement);
+      });
+    } else {
+      // Drop on the gap (insert before or insert after)
+      let ar;
+      let i: number;
+      loop(data, destKey, (item: TreeElement, index: number, arr: any[]) => {
+        ar = arr;
+        i = index;
+      });
+
+      if (destPosition === -1) {
+        // @ts-ignore
+        ar.splice(i, 0, srcElement);
+      } else {
+        // @ts-ignore
+        ar.splice(i + 1, 0, srcElement);
+      }
+    }
+
+    this.setState({
+      treeData: data,
+    });
+
+    this.reorderElements(srcEl, destEl);
+  };
+
+  reorderElements = (src: ElementState, dest: ElementState) => {
```

Comment:
a little hard to tell without digging more... is it possible to move the element manipulation *out* of the react context?  It would be nice if the logic to handle the UI were in the `.tsx` file, and the element manipulation were elsewhere (maybe in utils.ts) or in the scene directly